### PR TITLE
340 multiple vpc orig code place

### PR DESCRIPTION
**Pull request by ShiriMoran**

### CODE REVIEW

**Review by haim-kermany — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/output.go [hunk 1]
```diff
@@ -106,6 +106,7 @@ func NewOutputGenerator(c1, c2 map[string]*VPCConfig, grouping bool, uc OutputUs
 			}
 		}
 	}
+	unifyMultiVPC(c1, res.nodesConn, res.subnetsConn)
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/output.go]: done

**Review by haim-kermany — APPROVED** [review]

**Review by adisos — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/unifyGrouping.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+// unifyMultiVPC unifies multi-vpc graph for endpoints and subnets connectivity s.t.
+// each node appears once across multi-vpcs this is relevant only for DRAWIO and ARCHDRAWIO
+// in which there is a multivpc presentation
+// todo: this is actually required only for drawio, which is not accessible from the current context
+func unifyMultiVPC(config1 map[string]*VPCConfig, nodesConn map[string]*VPCConnectivity,
+	subnetsConn map[string]*VPCsubnetConnectivity) {
+	groupedEndpointsElemsMap := map[string]*groupedEndpointsElems{}
+	groupedExternalNodesMap := map[string]*groupedExternalNodes{}
+	for vpcName := range config1 {
+		if VPCconnectivity, ok := nodesConn[vpcName]; ok {
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/unifyGrouping.go:11]: right. done

File: pkg/vpcmodel/unifyGrouping.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+					unifiedGroupedConnLines(VPCconnectivity.GroupedConnectivity.GroupedLines,
+						groupedEndpointsElemsMap, groupedExternalNodesMap, true)
+			}
+		}
+		if subnetConnectivity, ok := subnetsConn[vpcName]; ok {
+			if len(subnetConnectivity.GroupedConnectivity.GroupedLines) > 0 {
+				subnetConnectivity.GroupedConnectivity.GroupedLines =
+					unifiedGroupedConnLines(subnetConnectivity.GroupedConnectivity.GroupedLines,
+						groupedEndpointsElemsMap, groupedExternalNodesMap, true)
+			}
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/unifyGrouping.go:23]: right, but after using `OutputUseCase`  I'm not sure its required.  The code duplication is only the assignment into func call return val

File: pkg/vpcmodel/grouping.go [hunk 4]
```diff
@@ -438,37 +439,55 @@ func (g *GroupConnLines) groupInternalSrcOrDst(srcGrouping, groupVsi bool) {
...
+// Go over the grouping result and set groups s.t. all semantically equiv groups have a unified reference.
+// this is required for multivpc's context and at the end of the grouping in a single vpc context
+// the former is required since each vpc analysis and grouping is done separately
+// the latter is required due to the functionality treating self loops as don't cares - extendGroupingSelfLoops
 // in which both srcs and dsts are manipulated  but *GroupConnLines is not familiar
 // within the extendGroupingSelfLoops context and thus can not be done there smoothly
-func (g *GroupConnLines) unifiedGroupedConnLines(oldConnLines []*groupedConnLine) []*groupedConnLine {
+func unifiedGroupedConnLines(oldConnLines []*groupedConnLine,
+	groupedEndpointsElemsMap map[string]*groupedEndpointsElems,
+	groupedExternalNodesMap map[string]*groupedExternalNodes,
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:453]: right. done

File: pkg/vpcmodel/unifyGrouping.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	}
+}
+
+// unifyMultiVPC unifies multi-vpc graph for endpoints and subnets connectivity s.t.
+// each node appears once across multi-vpcs this is relevant only for DRAWIO
+// in which there is a multivpc presentation
+func unifyMultiVPC(config1 map[string]*VPCConfig, nodesConn map[string]*VPCConnectivity,
+	subnetsConn map[string]*VPCsubnetConnectivity, uc OutputUseCase) {
+	cacheGrouped := initCacheGroupedElem()
+	for vpcName := range config1 {
```
Comment (adisos) [inline pkg/vpcmodel/unifyGrouping.go:21]: The keys in the map `config1 map[string]*VPCConfig` are VPC UIDs and not names (as documented in `parser.go` function `VPCConfigsFromResources()` ), so better rename var `vpcName` to `vpcUID`

File: pkg/vpcmodel/grouping.go [hunk 6]
```diff
@@ -68,43 +68,47 @@ func newGroupingConnections() *groupingConnections {
...
 	res = &GroupConnLines{config: c, nodesConn: v,
-		srcToDst:                 newGroupingConnections(),
-		dstToSrc:                 newGroupingConnections(),
-		groupedEndpointsElemsMap: map[string]*groupedEndpointsElems{},
-		groupedExternalNodesMap:  map[string]*groupedExternalNodes{}}
+		srcToDst: newGroupingConnections(),
+		dstToSrc: newGroupingConnections(),
+		cacheGrouped: cacheGroupedElements{
+			map[string]*groupedEndpointsElems{},
+			map[string]*groupedExternalNodes{}}}
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:75]: change the type of `cacheGrouped cacheGroupedElements` within `GroupConnLines` to be `*cacheGroupedElements`, and then use `initCacheGroupedElem()` here as well.

File: pkg/vpcmodel/grouping.go [hunk 7]
```diff
@@ -187,27 +190,28 @@ func (g *groupedExternalNodes) Name() string {
 
 // given a groupedEndpointsElems returns an equiv item from groupedEndpointsElemsMap if exists,
 // or adds it to groupedEndpointsElemsMap if such an item does not exist
-func (g *GroupConnLines) getGroupedEndpointsElems(grouped groupedEndpointsElems) *groupedEndpointsElems {
+func getGroupedEndpointsElems(grouped groupedEndpointsElems,
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:193]: This func should receive a reference to the cache, and use a cache method to check existence of element.
The cache-related functionality should be encapsulated by the cache type.

File: pkg/vpcmodel/grouping.go [hunk 8]
```diff
@@ -187,27 +190,28 @@ func (g *groupedExternalNodes) Name() string {
...
 		return existingGrouped
 	}
-	g.groupedEndpointsElemsMap[grouped.Name()] = &grouped
+	groupedEndpointsElemsMap[grouped.Name()] = &grouped
 	return &grouped
 }
 
 // same as the previous function, for groupedExternalNodesMap
-func (g *GroupConnLines) getGroupedExternalNodes(grouped groupedExternalNodes) *groupedExternalNodes {
+func getGroupedExternalNodes(grouped groupedExternalNodes,
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:205]: same here

**Review by adisos — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/grouping.go [hunk 9]
```diff
@@ -438,37 +442,50 @@ func (g *GroupConnLines) groupInternalSrcOrDst(srcGrouping, groupVsi bool) {
...
-// Go over the grouping result and make sure all groups have a unified reference.
-// this is required due to the functionality treating self loops as don't cares - extendGroupingSelfLoops
+// Go over the grouping result and set groups s.t. all semantically equiv groups have a unified reference.
+// this is required for multivpc's context and at the end of the grouping in a single vpc context
+// the former is required since each vpc analysis and grouping is done separately
+// the latter is required due to the functionality treating self loops as don't cares - extendGroupingSelfLoops
 // in which both srcs and dsts are manipulated  but *GroupConnLines is not familiar
 // within the extendGroupingSelfLoops context and thus can not be done there smoothly
-func (g *GroupConnLines) unifiedGroupedConnLines(oldConnLines []*groupedConnLine) []*groupedConnLine {
+func unifiedGroupedConnLines(oldConnLines []*groupedConnLine, cacheGrouped *cacheGroupedElements,
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:454]: consider adding unit tests, demonstrating why this functionality is required for the described cases (both for multivpc's context and at the end of the grouping in a single vpc context)

File: pkg/vpcmodel/unifyGrouping.go (hunk 5 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/unifyGrouping.go:21]: done

File: pkg/vpcmodel/grouping.go (hunk 6 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:75]: done

File: pkg/vpcmodel/grouping.go (hunk 7 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:193]: done

File: pkg/vpcmodel/grouping.go (hunk 9 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:425]: The effect of this unification is only in `drawio`  - the fixes originated in Haim's request. Its more natural to test it there. I can only test, or try to test, cumbersome side effects of the unification. 
Perhaps we can open an issue for this and see what it the best way to test it.

File: pkg/vpcmodel/grouping.go [hunk 10]
```diff
@@ -67,44 +67,44 @@ func newGroupingConnections() *groupingConnections {
 
 func newGroupConnLines(c *VPCConfig, v *VPCConnectivity,
 	grouping bool) (res *GroupConnLines, err error) {
+	initCacheGrouped := initCacheGroupedElem()
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:70]:
```suggestion
```

File: pkg/vpcmodel/grouping.go [hunk 11]
```diff
@@ -67,44 +67,44 @@ func newGroupingConnections() *groupingConnections {
...
 	res = &GroupConnLines{config: c, nodesConn: v,
-		srcToDst:                 newGroupingConnections(),
-		dstToSrc:                 newGroupingConnections(),
-		groupedEndpointsElemsMap: map[string]*groupedEndpointsElems{},
-		groupedExternalNodesMap:  map[string]*groupedExternalNodes{}}
+		srcToDst:     newGroupingConnections(),
+		dstToSrc:     newGroupingConnections(),
+		cacheGrouped: initCacheGrouped}
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:74]:
```suggestion
		cacheGrouped: initCacheGroupedElem()}
```

Comment (adisos) [reply pkg/vpcmodel/grouping.go:74]: and better rename `initCacheGroupedElem` to `newCacheGroupedElem`

File: pkg/vpcmodel/grouping.go [hunk 12]
```diff
@@ -67,44 +67,44 @@ func newGroupingConnections() *groupingConnections {
...
 	err = res.computeGrouping(true, grouping)
 	return res, err
 }
 
 func newGroupConnLinesSubnetConnectivity(c *VPCConfig, s *VPCsubnetConnectivity,
 	grouping bool) (res *GroupConnLines, err error) {
+	initCacheGrouped := initCacheGroupedElem()
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:81]: same here and in the other functions below

File: pkg/vpcmodel/grouping.go [hunk 13]
```diff
@@ -575,3 +564,62 @@ func (details *srcDstDetails) explanationEncode(c *VPCConfig) string {
...
+	grouped groupedEndpointsElems) *groupedEndpointsElems {
+	// since the endpoints (vsis/subnets) are sorted before printed, grouped.Name() will be identical
+	// to equiv groupedEndpointsElems
+	if existingGrouped, ok := cachedGrouped.groupedEndpointsElemsMap[grouped.Name()]; ok {
+		return existingGrouped
+	}
+	return nil
+}
+
+// gets pointer of an element semantically equiv to grouped in cachedGrouped.groupedEndpointsElemsMap
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:583]: move this part with all methods for `cacheGroupedElements` to `unifyGrouping.go` where this type is defined

File: pkg/vpcmodel/grouping_test.go [hunk 14]
```diff
@@ -146,10 +146,10 @@ func newVPCConfigTest2() (*VPCConfig, *VPCConnectivity) {
 // Test simple grouping of 2 conn lines with common src+conn, with dest as external ip ranges
 // thus, expecting to be merged to one line with dest element of both ranges together
 func TestGroupingPhase1(t *testing.T) {
+	initCacheGrouped := initCacheGroupedElem()
```
Comment (adisos) [inline pkg/vpcmodel/grouping_test.go:149]:
```suggestion
```

File: pkg/vpcmodel/grouping_test.go [hunk 15]
```diff
@@ -146,10 +146,10 @@ func newVPCConfigTest2() (*VPCConfig, *VPCConnectivity) {
...
 	c, v := newVPCConfigTest1()
 	res := &GroupConnLines{config: c, nodesConn: v, srcToDst: newGroupingConnections(), dstToSrc: newGroupingConnections(),
-		groupedEndpointsElemsMap: make(map[string]*groupedEndpointsElems),
-		groupedExternalNodesMap:  make(map[string]*groupedExternalNodes)}
+		cacheGrouped: initCacheGrouped}
```
Comment (adisos) [inline pkg/vpcmodel/grouping_test.go:152]:
```suggestion
		cacheGrouped: initCacheGroupedElem()}
```

Comment (adisos) [reply pkg/vpcmodel/grouping_test.go:152]: and in the functions below

**Review by adisos — CHANGES_REQUESTED** [review]
few small comments

File: pkg/vpcmodel/grouping.go (hunk 9 above)
Comment (adisos) [reply pkg/vpcmodel/grouping.go:425]: ok, consider open a separate issue for this

Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:425]: https://github.com/np-guard/vpc-network-config-analyzer/issues/429

File: pkg/vpcmodel/grouping.go (hunk 12 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go:80]: right

File: pkg/vpcmodel/grouping_test.go (hunk 14 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping_test.go]: done

File: pkg/vpcmodel/grouping_test.go (hunk 15 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/grouping_test.go]: done

File: pkg/vpcmodel/grouping.go [hunk 16]
```diff
@@ -77,33 +77,30 @@ func newGroupConnLines(c *VPCConfig, v *VPCConnectivity,
 
 func newGroupConnLinesSubnetConnectivity(c *VPCConfig, s *VPCsubnetConnectivity,
 	grouping bool) (res *GroupConnLines, err error) {
-	initCacheGrouped := initCacheGroupedElem()
 	res = &GroupConnLines{config: c, subnetsConn: s,
 		srcToDst:     newGroupingConnections(),
 		dstToSrc:     newGroupingConnections(),
-		cacheGrouped: initCacheGrouped}
+		cacheGrouped: initCacheGroupedElem()}
```
Comment (adisos) [inline pkg/vpcmodel/grouping.go:83]: rename to `newCacheGroupedElem`, to be consistent

File: pkg/vpcmodel/grouping.go (hunk 13 above)
Comment (adisos) [reply pkg/vpcmodel/grouping.go:579]: did you push this commit? I don't see any change

Comment (ShiriMoran) [reply pkg/vpcmodel/grouping.go]: oops. Pushed now

**Review by adisos — APPROVED** [review]

**Pull request merged**